File: QuickSort.java
import java.util.Random;

public class QuickSort {

    // Method to perform quick sort
    public void quickSort(int[] arr, int low, int high) {
        if (low < high) {
            // Partition the array
            int pi = partition(arr, low, high);

            // Recursively sort elements before and after partition
            quickSort(arr, low, pi - 1);
            quickSort(arr, pi + 1, high);
        }
    }

    // Method to partition the array
    public int partition(int[] arr, int low, int high) {
        int pivot = arr[high]; // Pivot element
        int i = (low - 1); // Index of the smaller element

        for (int j = low; j < high; j++) {
            // If the current element is smaller than the pivot
            if (arr[j] < pivot) {
                i++;

                // Swap arr[i] and arr[j]
                int temp = arr[i];
                arr[i] = arr[j];
                arr[j] = temp;
            }
        }

        // Swap arr[i + 1] and arr[high] (or pivot)
        int temp = arr[i + 1];
        arr[i + 1] = arr[high];
        arr[high] = temp;

        return i + 1; // Return the partitioning index
    }

    // Method to fill the array with random numbers
    public void fillArrayWithRandomNumbers(int[] array) {
        Random random = new Random();
        for (int i = 0; i < array.length; i++) {
            array[i] = random.nextInt(1000000); // Random numbers between 0 and 999999
        }
    }

    public static void main(String[] args) {
        int arraySize = 300000; // Number of elements
        int[] randomNumbers = new int[arraySize];
        
        QuickSort obj = new QuickSort();
        obj.fillArrayWithRandomNumbers(randomNumbers);

        long start = System.nanoTime();
        obj.quickSort(randomNumbers, 0, randomNumbers.length - 1);
        long end = System.nanoTime();

        long elapsedTimeInNanoseconds = end - start;
        double elapsedTimeInMilliseconds = elapsedTimeInNanoseconds / 1_000_000.0;

        // Print the result and time taken
        System.out.println("Array sorted using quick sort.");
        System.out.println("Time taken for quick sort: " + elapsedTimeInMilliseconds + " milliseconds");
    }
}

File: Python-Files/QuickSort.py
class QuickSort:
    def sort(self, arr):
        self._quick_sort_iterative(arr, 0, len(arr) - 1)

    def _quick_sort_iterative(self, arr, low, high):
        stack = [(low, high)]
        while stack:
            low, high = stack.pop()
            if low < high:
                pivot_index = self._partition(arr, low, high)
                stack.append((low, pivot_index - 1))
                stack.append((pivot_index + 1, high))

    def _partition(self, arr, low, high):
        pivot = arr[high]
        i = low - 1
        for j in range(low, high):
            if arr[j] < pivot:
                i += 1
                arr[i], arr[j] = arr[j], arr[i]
        arr[i + 1], arr[high] = arr[high], arr[i + 1]
        return i + 1


if __name__ == "__main__":
    array = [38, 27, 43, 3, 9, 82, 10]
    sorter = QuickSort()
    sorter.sort(array)
    print("Sorted array using Quick Sort:", array)
